src/ray/gcs/asio.cc

```diff
@@ -16,9 +20,21 @@ RedisAsioClient::RedisAsioClient(boost::asio::io_service &io_service,
   // gives access to c->fd
   redisContext *c = &(async_context->c);
 
+#ifdef _WIN32
+  SOCKET sock = SOCKET_ERROR;
+  WSAPROTOCOL_INFO pi;
+  if (WSADuplicateSocket(_get_osfhandle(c->fd), GetCurrentProcessId(), &pi) == 0) {
+    DWORD flag = WSA_FLAG_OVERLAPPED;
+    sock = WSASocket(pi.iAddressFamily, pi.iSocketType, pi.iProtocol, &pi, 0, flag);
+  }
+  boost::asio::ip::tcp::socket::native_handle_type handle(sock);
+#else
+  boost::asio::ip::tcp::socket::native_handle_type handle(dup(c->fd));
```

Comment:
why does the fd need to be duped here?

Reply:
It's because Boost.Asio socket appears to take ownership of the file descriptor. (I think it's a bug in [hiredis-boostasio-adapter](https://github.com/ryangraham/hiredis-boostasio-adapter/blob/32cdf7172474d919f15f217a15ecab7b630e73d7/boostasio.cpp#L17).)

I tested it with the following, which only succeeds if `sock` is moved outside the braces:

	#include <errno.h>
	#include <stdlib.h>

	#include <boost/asio.hpp>

	int main()
	{
		boost::asio::io_service io_service;
		boost::asio::detail::socket_type s = -1;
		{
			boost::asio::ip::tcp::socket sock(io_service);
			s = socket(PF_INET, SOCK_STREAM, IPPROTO_TCP);
			sock.assign(boost::asio::ip::tcp::v4(), s);
		}
		int val;
		int len = sizeof(val);
		return getsockopt(s, SOL_SOCKET, SO_ACCEPTCONN, (char *)&val, &len);
	}

Reply:
Ok, makes sense. I guess in the worst case it leaks an fd in a code path that only runs once. I agree that it looks like it takes ownership, also based on this: https://stackoverflow.com/questions/2966323/boostasio-and-socket-ownership

Reply:
Yup, I thought the same thing.  
I just reported it upstream as well: https://github.com/ryangraham/hiredis-boostasio-adapter/issues/3